storagenode/piecestore/retain.go

```diff
@@ -0,0 +1,186 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package piecestore
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"storj.io/storj/pkg/bloomfilter"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/storagenode/pieces"
+)
+
+// RetainService queues and processes retain requests from satellites
+type RetainService struct {
+	log          *zap.Logger
+	retainStatus RetainStatus
+
+	mu     sync.Mutex
+	queued map[storj.NodeID]RetainRequest
+
+	reqChan      chan RetainRequest
+	sem          chan struct{}
+	emptyTrigger chan struct{}
+
+	store *pieces.Store
+}
+
+// RetainRequest contains all the info necessary to process a retain request
+type RetainRequest struct {
+	SatelliteID   storj.NodeID
+	CreatedBefore time.Time
+	Filter        *bloomfilter.Filter
+}
+
+// NewRetainService creates a new retain service
+func NewRetainService(log *zap.Logger, retainStatus RetainStatus, concurrentRetain int, store *pieces.Store) *RetainService {
+	return &RetainService{
+		log:          log,
+		retainStatus: retainStatus,
+		queued:       make(map[storj.NodeID]RetainRequest),
+		reqChan:      make(chan RetainRequest),
+		sem:          make(chan struct{}, concurrentRetain),
+		emptyTrigger: make(chan struct{}),
+		store:        store,
+	}
+}
+
+// QueueRetain adds a retain request to the queue
+func (s *RetainService) QueueRetain(req RetainRequest) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// only queue retain request if we do not already have one for this satellite
+	if _, ok := s.queued[req.SatelliteID]; !ok {
+		s.queued[req.SatelliteID] = req
+		go func() { s.reqChan <- req }()
+	}
+}
+
+// Run listens for queued retain requests and processes them as they come in
+func (s *RetainService) Run(ctx context.Context) error {
+	for {
+		// exit if context has been canceled. Otherwise, block until an item can be added to the semaphore
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case s.sem <- struct{}{}:
+		}
+
+		// get the next request
+		var req RetainRequest
+		select {
+		case req = <-s.reqChan:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+
+		// TODO make it possible to sync with this goroutine
+		go func(ctx context.Context, req RetainRequest) {
+			err := s.retainPieces(ctx, req)
+			if err != nil {
+				s.log.Error("retain error", zap.Error(err))
+			}
+			s.mu.Lock()
+			delete(s.queued, req.SatelliteID)
+			s.mu.Unlock()
+
+			if len(s.queued) == 0 {
```

Comment:
If we don't protect len(s.queued) with the mutex, do we risk missing an incoming Retain() during a Wait()... it doesn't seem like that's an issue or not

Reply:
I'm not sure. I can protect this part with the mutex as well.